compiler/cpp/src/reader/Reader.cpp

```diff
@@ -0,0 +1,53 @@
+//
+// Created by Marc on 23.05.2021.
+//
+
+#include "Reader.h"
+
+Reader::Reader(std::string fileInput, unsigned int maxLookahead):
+    fileInput(std::move(fileInput)), maxLookahead(maxLookahead) {
+    advance(); // Load first char into buffer
+}
+
+void Reader::advance() {
+    // Return EOF, when string ends
+    if (inputStringPos >= fileInput.length()) {
+        curChar = EOF;
+        return;
+    }
+    // Last char was a line break?
+    if (curChar == '\n') {
+        lineNum++;
+        colNum = 0;
+    }
+    // Go to next char
+    curChar = (unsigned char) fileInput[inputStringPos];
+    inputStringPos++;
+    colNum++;
+}
+
+void Reader::expect(int input) {
+    if (curChar != input) throw UnexpectedCharException((char) input, (char) curChar, lineNum, colNum);
+    advance();
+}
+
+void Reader::expectMultiple(const std::string& input) {
+    for (char const &c: input) expect(c);
+}
+
+int Reader::getLookahead() const {
+    return curChar;
+}
+
+std::string Reader::getMaxLookahead() {
+    int length = std::min((int) fileInput.length() - inputStringPos, (int) maxLookahead);
+    return fileInput.substr(inputStringPos -1, length);
```

Comment:
why `inputStringPos - 1`?

Reply:
That is because of the common "Index problem". For us programmers, the inputString starts at col 0, but the error message for the user has to be Line 1, Col 1.

Reply:
@marcauberer Yeah, I was just wondering since you initialize `inputStringPos` with 0 in the header file [here](https://github.com/compose-generator/ccom/pull/25/files/9eaf8dffdce3b29fc7646a2d0281c4f8b2634274#diff-2f43142f88a40da9e24dfd968b4fd29fa62833361e9262481b990a4adf327d72R20)

Reply:
Ok. I will take a look if we could solve this more elegantly. Can we first merge pr #28 into this one, so I can test it more easily?

Reply:
@marcauberer Yeah, I will go ahead and review #28 then.